adam-core/src/main/scala/org/bdgenomics/adam/rdd/settheory/ShuffleRegionJoin.scala

```diff
@@ -0,0 +1,334 @@
+ * @tparam U The type of the right records.
+ * @tparam RT The resulting type of the left after the join.
+ * @tparam RU The resulting type of the right after the join.
+ */
+sealed abstract class ShuffleRegionJoin[T: ClassTag, U: ClassTag, RT, RU]
+    extends SetTheoryBetweenCollections[T, U, RT, RU] with SetTheoryPrimitive {
+
+  override protected def condition(firstRegion: ReferenceRegion,
+                                   secondRegion: ReferenceRegion,
+                                   distanceThreshold: Long = 0L): Boolean = {
+
+    firstRegion.isNearby(secondRegion,
+      distanceThreshold,
+      requireStranded = false)
+  }
+
+  override protected def pruneCacheCondition(cachedRegion: ReferenceRegion,
+                                             to: ReferenceRegion): Boolean = {
+
+    cachedRegion.compareTo(to) < 0 && !cachedRegion.covers(to)
+  }
+
+  override protected def advanceCacheCondition(candidateRegion: ReferenceRegion,
+                                               until: ReferenceRegion): Boolean = {
+
+    candidateRegion.compareTo(until) < 0 || candidateRegion.covers(until)
+  }
+
+  override protected def prepare(): (RDD[(ReferenceRegion, T)], RDD[(ReferenceRegion, U)]) = {
+    
+    val (preparedLeft, destinationPartitionMap) = {
+      if(optPartitionMap.isDefined) {
+        (leftRdd, optPartitionMap.get.map(_.get))
+      } else {
+        val sortedLeft = leftRdd.sortByKey()
+        val partitionMap =
+          sortedLeft.mapPartitions(getRegionBoundsFromPartition).collect
+        (sortedLeft, partitionMap.map(_.get))
+      }
+    }
+
+    val adjustedPartitionMapWithIndex =
+    // the zipWithIndex gives us the destination partition ID
+      destinationPartitionMap.zipWithIndex.map(g => {
+        val (firstRegion, secondRegion, index) = (g._1._1, g._1._2, g._2)
+        // in the case where we span multiple referenceNames using
+        // IntervalArray.get with requireOverlap set to false will assign all
+        // the remaining regions to this partition, in addition to all the
+        // regions up to the start of the next partition.
+        if (firstRegion.referenceName != secondRegion.referenceName) {
+
+          // the first region is enough to represent the partition for
+          // IntervalArray.get.
+          (firstRegion, index)
+        } else {
+          // otherwise we just have the ReferenceRegion span from partition
+          // lower bound to upper bound.
+          // We cannot use the firstRegion bounds here because we may end up
+          // dropping data if it doesn't map anywhere.
+          (ReferenceRegion(
+            firstRegion.referenceName,
+            firstRegion.start,
+            secondRegion.end),
+            index)
+        }
+      })
+
+    // convert to an IntervalArray for fast range query
+    val partitionMapIntervals = IntervalArray(
+      adjustedPartitionMapWithIndex,
+      adjustedPartitionMapWithIndex.maxBy(_._1.width)._1.width,
```

Comment:
This was an attempted optimization. Rather than iterating over the entire list twice, we do it once with `maxBy`. There is certainly little effect because the number of partitions will presumably be relatively small. I'll go ahead and change it.